From cdf5b8190a931c9cff7b4eefd4e19f6eab002408 Mon Sep 17 00:00:00 2001
From: "Christopher J. Morrone" <morrone2@llnl.gov>
Date: Thu, 21 Apr 2016 02:39:36 +0000
Subject: [PATCH 11/18] Allow slaves to substantiate in parallel

BuildRequestDistributor's _maybeStartBuildsOnBuilder() method is
decorated with @defer.inlineCallbacks.  This means that when it
uses "yield bldr.maybeStartBuild", the entire slave substantiation
process is sequentialized at that point.

To enable parallel slave substantiation, we need to move the error
handling code into a callback function that can be added to the
Deferred that is returned by bldr.maybeStartBuild().

Fixes #4

Signed-off-by: Christopher J. Morrone <morrone2@llnl.gov>
---
 master/buildbot/process/buildrequestdistributor.py | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/master/buildbot/process/buildrequestdistributor.py b/master/buildbot/process/buildrequestdistributor.py
index 2848510ab..a6d88cee0 100644
--- a/master/buildbot/process/buildrequestdistributor.py
+++ b/master/buildbot/process/buildrequestdistributor.py
@@ -531,14 +531,15 @@ class BuildRequestDistributor(service.Service):
                 bc = self.createBuildChooser(bldr, self.master)
                 continue
 
-            buildStarted = yield bldr.maybeStartBuild(slave, breqs)
-
-            if not buildStarted:
-                yield self.master.db.buildrequests.unclaimBuildRequests(brids)
-
-                # and try starting builds again.  If we still have a working slave,
-                # then this may re-claim the same buildrequests
-                self.botmaster.maybeStartBuildsForBuilder(self.name)
+            d = bldr.maybeStartBuild(slave, breqs)
+            @defer.inlineCallbacks
+            def checkBuildStart(buildStarted, slavename, buildername, brids):
+                if not buildStarted:
+                    yield self.master.db.buildrequests.unclaimBuildRequests(brids)
+                    # and try starting builds again.  If we still have a working slave,
+                    # then this may re-claim the same buildrequests
+                    self.botmaster.maybeStartBuildsForBuilder(buildername)
+            d.addCallback(checkBuildStart, slave.slave.slavename, bldr.name, brids)
 
     def createBuildChooser(self, bldr, master):
         # just instantiate the build chooser requested
-- 
2.13.6

